androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/NotificationPermissionRequestActivity.java

```diff
@@ -0,0 +1,123 @@
+import android.content.Context;
+import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.os.Bundle;
+import android.os.Message;
+import android.os.Messenger;
+import android.os.RemoteException;
+
+import android.util.Log;
+import androidx.core.app.ActivityCompat;
+
+/**
+ * A simple transparent activity for requesting the notification permission. On either approve or disapprove, this will
+ * send the result via the {@link Messenger} provided with the intent, and then finish.
+ */
+public class NotificationPermissionRequestActivity extends Activity {
+    private static final String TAG = "Notifications";
+
+    static final String KEY_PERMISSION_STATUS = "permissionStatus";
+
+    // TODO: Use Manifest.permission.POST_NOTIFICATIONS when it is released.
+    private static final String PERMISSION_POST_NOTIFICATIONS = "android.permission.POST_NOTIFICATIONS";
+
+    // TODO: Use Build.VERSION_CODES when it is released.
+    private static final int VERSION_T = 33;
+
+    private static final String EXTRA_NOTIFICATION_CHANNEL_NAME = "notificationChannelName";
+    private static final String EXTRA_MESSENGER = "messenger";
+
+    private String mChannelName;
+    private Messenger mMessenger;
+
+    /**
+     * Creates a {@link PendingIntent} for launching this activity to request the notification permission. It is mutable
+     * so that a messenger extra can be added for returning the permission request result.
+     */
+    public static PendingIntent createPermissionRequestPendingIntent(Context context, String channelName) {
+        Intent intent = new Intent(context.getApplicationContext(), NotificationPermissionRequestActivity.class);
+        intent.putExtra(EXTRA_NOTIFICATION_CHANNEL_NAME, channelName);
+        // Starting with Build.VERSION_CODES.S it is required to explicitly specify the mutability of PendingIntents.
+        int flags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S ? PendingIntent.FLAG_MUTABLE : 0;
+        return PendingIntent.getActivity(context.getApplicationContext(),0, intent, flags);
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        mChannelName = getIntent().getStringExtra(EXTRA_NOTIFICATION_CHANNEL_NAME);
+        mMessenger = getIntent().getParcelableExtra(EXTRA_MESSENGER);
+        if (mChannelName == null || mMessenger == null) {
+            Log.w(TAG, "Finishing because no channel name or messenger for returning the result was provided.");
+            finish();
+            return;
+        }
+
+        // When running on T or greater, with the app targeting less than T, creating a channel for the first time will
+        // trigger the permission dialog.
+        if (Build.VERSION.SDK_INT >= VERSION_T && getApplicationContext().getApplicationInfo().targetSdkVersion < VERSION_T) {
+            NotificationUtils.createNotificationChannel(this, mChannelName);
```

Comment:
Do we want a `return` here?

Reply:
Not really. If we call requestPermissions and asynchronously end up in onRequestPermissionsResult, the permission dialog can be shown and processed in between, whether it was triggered by requestPermissions or createNotificationChannel. This gives the user the chance to grant permission, and then we can read that in onRequestPermissionsResult and send the result back via the messenger.